src/san_bump.c

```diff
@@ -0,0 +1,121 @@
+#include "jemalloc/internal/jemalloc_preamble.h"
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+#include "jemalloc/internal/san_bump.h"
+#include "jemalloc/internal/pac.h"
+#include "jemalloc/internal/san.h"
+#include "jemalloc/internal/ehooks.h"
+#include "jemalloc/internal/edata_cache.h"
+
+static bool
+san_bump_alloc_grow_locked(tsdn_t *tsdn, san_bump_alloc_t *sba, pac_t *pac,
+    ehooks_t *ehooks, size_t alloc_size_min);
+
+bool
+san_bump_alloc_init(san_bump_alloc_t* sba) {
+	bool err = malloc_mutex_init(&sba->mtx, "sanitizer_bump_allocator",
+	    WITNESS_RANK_SAN_BUMP_ALLOC, malloc_mutex_rank_exclusive);
+	if (err) {
+		return true;
+	}
+	exp_grow_init(&sba->exp_grow);
+	edata_addr_set(&sba->curr_reg, NULL);
+
+	return false;
+}
+
+edata_t *
+san_bump_alloc(tsdn_t *tsdn, san_bump_alloc_t* sba, pac_t *pac,
+    ehooks_t *ehooks, size_t size, size_t alignment, bool zero) {
+	assert(opt_retain);
+	assert(alignment <= PAGE);
+
+	size_t guarded_size = san_one_side_guarded_sz(size);
+
+	malloc_mutex_lock(tsdn, &sba->mtx);
+
+	bool grew_retained = false;
+	if (edata_addr_get(&sba->curr_reg) == NULL ||
+	    edata_size_get(&sba->curr_reg) < guarded_size) {
+		bool err = san_bump_alloc_grow_locked(tsdn, sba, pac, ehooks,
+		    guarded_size);
+		if (err) {
+			goto label_err;
+		}
+		grew_retained = true;
+	}
+	edata_t *edata = edata_cache_get(tsdn, pac->edata_cache);
+	if (edata == NULL) {
+		goto label_err;
+	}
+
+	edata_init(edata, edata_arena_ind_get(&sba->curr_reg),
+	    edata_addr_get(&sba->curr_reg), guarded_size,
+	    false, SC_NSIZES, edata_sn_get(&sba->curr_reg), extent_state_active,
+	    edata_zeroed_get(&sba->curr_reg),
+	    edata_committed_get(&sba->curr_reg), EXTENT_PAI_PAC,
+	    grew_retained ? EXTENT_IS_HEAD : EXTENT_NOT_HEAD);
+
+	san_guard_pages(tsdn, ehooks, edata, pac->emap, /* left */ false,
+	    /* right */ true, /* remap */ false);
+	extent_register_no_gdump_add(tsdn, pac, edata);
+
+	if (!edata_committed_get(edata)) {
+		if (extent_commit_wrapper(tsdn, ehooks, edata, 0,
+		    edata_size_get(edata), true)) {
+			extent_record(tsdn, pac, ehooks, &pac->ecache_retained,
+			    edata);
+			goto label_err;
+		}
+		edata_committed_set(edata, true);
+	}
+
+	if (zero && !edata_zeroed_get(edata)) {
```

Comment:
Just realized, don't we also need a `exp_grow_size_commit` for san_bump_alloc_grow (maybe rename it to san_bump_grow since it doesn't do the actual alloc part)?

For the commit error handling, I think we can leave the grow_retained part unchanged (i.e. not using the new utility created for this). That function can handle an arbitrarily large allocation, so the chance of commit failure is real. Avoiding exp_grow_size_commit in such failure case also has its point.

While the bump alloc here is for small alloc only, it's probably fine that we don't even use the exp_grow.